### ISSUE TITLE
Define metric extension mechanism

### ISSUE THREAD

**Issue by lmolkova** [open-telemetry/semantic-conventions#760]
We have generic metrics (such as `http.server.request.duration` or `messaging.publish,duration`).
We also have precedents extending such metrics.
E.g. .NET semconv extends [`http.server.request.duration`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/dotnet/dotnet-http-metrics.md#metric-httpserverrequestduration) and adds .NET-specific attribute on it.

In case of messaging, in #687 we're adding [system-specific attributes](https://github.com/open-telemetry/semantic-conventions/blob/caf936e16d07282e036db524e70bc196ed262a3e/docs/messaging/kafka.md#metric-messagingpublishmessages)

On Kafka, we'd like to report partition id.  As a result, if we proceed with #687 we'll have a common set of metrics, but each individual messaging system would be able to add its own attributes.

We have other similar cases:
- connection-level metrics (we have overlapping ones for DB and .NET http connections).
- DB generic call duration and custom attributes added by extensions

----------

**Problem**:  We don't have a uniform solution on how to approach extension. 

Possible options are:

**Option 1** Create different metric for each db/messaging/rpc system) with their own set of attributes
E.g.
- messaging.kafka.send.duration 
- messaging.rabbitmq.publish.duration
- ....

Pros: 
- can follow system-specific terminology

Cons:
- no common dashboards/alerts 
- hard to make metrics consistent.  Possible mitigation could be that extension mechanism that allows to update metric name, but inherits instrument, unit, and attributes.

**Option 2:** Have common metric, add attributes to it in the extension.
As a result, we'll have `messaging.publish.duration` with some common attributes.
- kafka metric would also have `messaging.system="kafka", messaging.kafka.partition_id="foo"` attributes
- rabbitmq would have `messaging.system="rabbitmq", messaging.rabbitmq.routing_key="bar"` attributes

Pros: 
- can create generic dashboards/alerts which usually need to be adjusted for each system 
- according to @trask "it's the least bad option"

Cons:
- if I have multiple messaging systems, it could be overwhelming
- terminology differences
- sometimes we generalize things that are not the same and could end up with different concepts captured with one metric. This is not strictly related to extension mechanism.

----------

Regardless of the option we pick, we'd need to add some tooling to support extensibility. Related #592

**Comment by pyohannes** [comment]
The second option seems the more intuitive, and is how we extend spans with system-specific attributes.

However, for metrics this somehow breaks with the assumption of a stable set of attributes (and consequently number of time series) for metrics (see discussions in #722). Due to our current consensus around stability for metrics, we don't allow additional attributes to be added to metrics which would result in a split of existing time series, as this might break alerts for customers.

Regarding the proposed option (2) in this issue, let's assume a customer is using Kafka, obtains Kafka metrics (with both generic and Kafka-specific attributes), and defines alerts for those metrics. If this customer now also starts to use RabbitMQ and obtains RabbitMQ metrics, existing time series used for Kafka-related alerts might split and alerts work differently.

If we put up a high bar for adding additional attributes to stable metrics (which I'd like to question), this issue is likely also something we should take seriously.

**Comment by trask** [comment]
> let's assume a customer is using Kafka, obtains Kafka metrics (with both generic and Kafka-specific attributes), and defines alerts for those metrics. If this customer now also starts to use RabbitMQ and obtains RabbitMQ metrics, existing time series used for Kafka-related alerts might split and alerts work differently.

I don't think this would be considered breaking, since in this case the customer added a new instrumentation (instead of solely upgrading an existing instrumentation).

I do worry though that it would be considered breaking if Kafka instrumentation initially emits the stable generic metrics and then we add a (recommended) Kafka-specific attribute after stability.

**Comment by pyohannes** [comment]
See also this relevant comment: https://github.com/open-telemetry/semantic-conventions/issues/772#issuecomment-1966646924.

**Comment by lmolkova** [comment]
Based on the messaging SIG discussion last week (02/28):
- we need to decide whether generic dashboards/alerts are valuable enough. If they are, we should target option 2.
- we can still add attributes to metrics post-stability as opt-in (but should do the best effort to add everything important pre-stability)
- we need to build more confidence in the attributes we want to include on RabbitMQ and Kafka metrics 

There is still concern on systems that don't have semconv extensions.

E.g. we have `db|messaging.system = foo` for which we only declared a constant and it does not have any extensions.
We can declare stability for the `foo` enum member and then corresponding library instrumentation can go stable.
**Later on we might find out that some important attribute is missing on the generic metrics for `foo` and would need to add one which would be breaking at that point.**

Options to mitigate would be 1) opt-in attribute 2) new custom metric for foo 3) breaking change to foo, etc listed in #772

Options to minimize future breaking changes would be
- introduce more generic attributes for common things like partition-id, consumer-group, subscription-name. 
- update `messaging.destination.name` (or add a new attribute) to be even more granular one to include partition-id/consumer-group/subscription-name. E.g. in case of Azure messaging we have a notion of entity-path which looks like `{topic-name}/ConsumerGroups/{consumer-group}/Partitions/{partition-id}` . The downside is that producer and consumer destinations are different.

**Comment by lmolkova** [comment]
Throwing additional thoughts into this discussion:

when `*.system` attribute is defined, adding system-specific attributes seems reasonable.

But for metrics like `socket.connection.client.duration` having custom attributes would arguably be impossible. Those would be set in stone unless we introduce a discriminator.

**Comment by pyohannes** [comment]
> Option 2: Have a common metric, add custom attributes to it in the extensions

As discussed in the last workgroup meeting, this is the preferred option.

* For initial stability, we will provide conventions for generic metrics (sufficient for Kafka), with a RabbitMQ extension that defines additional specific attributes (for the routing key).
* Once stability is declared, those semantic conventions applicable for Kafka and RabbitMQ cannot be extended anymore.
* Additional messaging systems can define extensions to metrics and additional attributes, as long as system-specific conventions are not stable (as long as the corresponding value for `messaging.system` is not stable).
* Once system-specific semantic conventions are stable, not additional system-specific metric attributes can be defined.

@open-telemetry/semconv-messaging-approvers Do you think we can close this issue with that conclusion? A question in particular for @trask: Should we document this "stability requirements" somewhere as part of semantic conventions?
